_docs/implementation.md

```diff
@@ -0,0 +1,121 @@
+
+The goal is to provide a flexible and non-intrusive mechanism for instrumenting Go applications, particularly for use cases such as observability (e.g., OpenTelemetry), debugging, or performance profiling.
+
+# Core Principles
+## 1. Trampoline Code Injection
+Trampoline function calls are injected into the Target (lib-side) function, which ultimately jumps to the actual hook code via the function pointer Hook.
+
+```go
+func Target() {
+    Trampoline()
+    ....
+}
+
+func Trampoline() {
+    Hook()
+}
+
+//go:linkname Hook github.com/open-telemetry/opentelemetry-go-compile-instrumentation/sdk/hook.MyHook
+var Hook func()
+```
+The two-level indirection (trampoline call to hook) is a deliberate design choice that offers several key benefits:
+
+- Exception Handling: The trampoline catches panics and isolates exception handling, preventing them from affecting the target function or hook code.
+- Context Construction: The trampoline initializes and manages the necessary context before invoking the hook code.
+- Decoupling: The trampoline decouples the hook code from the target function, enabling flexibility and maintainability.
+- Debugging: The trampoline provides a centralized point for debugging and observability.
+
+
+## 2. Linkage via golinkname
+The `Hook` function is linked to the monitoring code using the `//go:linkname` directive. This allows us to dynamically associate the hook with the target function at compile time.
+
+# Implementation Details
+Go compile-time instrumentation is a two-phase process: the first phase involves updating dependencies, and the second phase focuses on building the project using a custom toolchain.
+
+## Phase 1: Adding Dependencies
+To prepare the project for instrumentation, necessary dependencies are integrated into the user's codebase. This step ensures that the required hook code is available during the build process.
+
+This can be done by creating or modifying files. For example:
+
+```go
+// otel_importer.go
+package main
+
+import _ "github.com/open-telemetry/opentelemetry-go-compile-instrumentation/sdk"
+```
+
+This import statement ensures that the SDK, which contains the hook code, is included in the project. The `_` prefix indicates that the package is imported only for its side effects (e.g., registration of hooks).
+
+After adding the dependency, `go mod tidy` is run to update the `go.mod` file. This step ensures that the dependency is properly recorded and downloaded into the module cache.
+By completing this phase, the project is now ready for the next step, where the instrumentation logic will be applied during the build process.
+
+## Phase 2: Building with Instrumentation
+The build process is invoked with a custom toolchain by specifying the `-toolexec=otel` flag
+
+```bash
+go build -toolexec=otel
+```
```

Comment:
I wonder if we should explicitly support folks specifying `-toolexec` "by hand" or if we should wrap `go build` entirely (and manage it behind the scenes). It can be useful to have a way to maintain state throughout the entire build, and that is made significantly easier if we control the root process...

I'm not entirely certain this is necessary, though...

Reply:
I prefer to support both `otel go build` and `go build -toolexec=otel`. Perhaps I should explicitly mention in the documentation that the motivation for supporting `go build -toolexec=otel` is that it allows us to directly append `-toolexec=otel` via the `GOFLAGS` environment variable, without having to modify the build command (`go build`).

Reply:
> @y1yang0 +1 to support both of the usage. Would you please add it to the doc?

Done